obs-studio-server/source/osn-source.cpp

```diff
@@ -734,6 +775,131 @@ void osn::Source::SetEnabled(
+		rval.push_back(ipc::value((uint64_t)ErrorCode::InvalidReference));
+		rval.push_back(ipc::value("Source reference is not valid."));
+		AUTO_DEBUG;
+		return;
+	}
+
+	obs_mouse_event event = {
+	    args[1].value_union.ui32,
+	    args[2].value_union.ui32,
+	    args[3].value_union.ui32,
+	};
+
+	obs_source_send_mouse_click(
+	    src, &event, args[4].value_union.ui32, args[5].value_union.i32, args[6].value_union.ui32);
+
+	AUTO_DEBUG;
+}
+
+void osn::Source::SendMouseMove(
+    void*                          data,
+    const int64_t                  id,
+    const std::vector<ipc::value>& args,
+    std::vector<ipc::value>&       rval)
+{
+	obs_source_t* src = osn::Source::Manager::GetInstance().find(args[0].value_union.ui64);
+
+	if (src == nullptr) {
+		rval.push_back(ipc::value((uint64_t)ErrorCode::InvalidReference));
+		rval.push_back(ipc::value("Source reference is not valid."));
+		AUTO_DEBUG;
+		return;
+	}
+
+	obs_mouse_event event = {args[1].value_union.ui32, args[2].value_union.ui32, args[3].value_union.ui32};
+
+	obs_source_send_mouse_move(src, &event, args[4].value_union.i32);
+
+	AUTO_DEBUG;
+}
+
+void osn::Source::SendMouseWheel(
+    void*                          data,
+    const int64_t                  id,
+    const std::vector<ipc::value>& args,
+    std::vector<ipc::value>&       rval)
+{
+	obs_source_t* src = osn::Source::Manager::GetInstance().find(args[0].value_union.ui64);
+
+	if (src == nullptr) {
+		rval.push_back(ipc::value((uint64_t)ErrorCode::InvalidReference));
+		rval.push_back(ipc::value("Source reference is not valid."));
+		AUTO_DEBUG;
+		return;
+	}
+
+	obs_mouse_event event = {
+	    args[1].value_union.ui32,
+	    args[2].value_union.ui32,
+	    args[3].value_union.ui32,
+	};
+
+	obs_source_send_mouse_wheel(src, &event, args[4].value_union.i32, args[5].value_union.i32);
+
+	AUTO_DEBUG;
+}
+
+void osn::Source::SendFocus(
+    void*                          data,
+    const int64_t                  id,
+    const std::vector<ipc::value>& args,
+    std::vector<ipc::value>&       rval)
+{
+	obs_source_t* src = osn::Source::Manager::GetInstance().find(args[0].value_union.ui64);
+
+	if (src == nullptr) {
+		rval.push_back(ipc::value((uint64_t)ErrorCode::InvalidReference));
+		rval.push_back(ipc::value("Source reference is not valid."));
+		AUTO_DEBUG;
+		return;
+	}
+
+	obs_source_send_focus(src, args[1].value_union.i32);
+
+	AUTO_DEBUG;
+}
+
+void osn::Source::SendKeyClick(
+    void*                          data,
+    const int64_t                  id,
+    const std::vector<ipc::value>& args,
+    std::vector<ipc::value>&       rval)
+{
+	obs_source_t* src = osn::Source::Manager::GetInstance().find(args[0].value_union.ui64);
+
+	if (src == nullptr) {
+		rval.push_back(ipc::value((uint64_t)ErrorCode::InvalidReference));
+		rval.push_back(ipc::value("Source reference is not valid."));
+		AUTO_DEBUG;
+		return;
+	}
+
+	char* text = new char[args[2].value_str.size()];
+	strcpy(text, args[2].value_str.c_str());
```

Comment:
isnt it need to be size()+1 to copy zero character at the end of the c string?
can we send "args[2].value_str.c_str()" without creating new string?

Reply:
No `obs_key_event` only takes a `char*` not `const char*`
Fixing the size, good catch